packages/flutter/lib/src/widgets/restoration_properties.dart

```diff
@@ -0,0 +1,250 @@
+
+  final T _defaultValue;
+
+  @override
+  T createDefaultValue() => _defaultValue;
+
+  @override
+  void didUpdateValue(T oldValue) {
+    assert(debugIsSerializableForRestoration(value));
+    notifyListeners();
+  }
+
+  @override
+  T fromPrimitives(Object serialized) {
+    return serialized as T;
+  }
+
+  @override
+  Object toPrimitives() {
+    return value;
+  }
+}
+
+/// A [RestorableProperty] that knows how to store and restore a [num].
+///
+/// {@template flutter.widgets.restoration.primitivevalue}
+/// The current [value] of this property is stored in the restoration data.
+/// During state restoration the property is restored to the value it had when
+/// the restoration data it is getting restored from was collected.
+///
+/// If no restoration data is available, [value] is initialized to the
+/// `defaultValue` given in the constructor.
+/// {@endtemplate}
+///
+/// See also:
+///
+///  * [RestorableDouble], which can store and restore a [double].
+///  * [RestorableInt], which can store and restore an [int].
+class RestorableNum<T extends num> extends _RestorablePrimitiveValue<T> {
+  /// Creates a [RestorableNum].
+  ///
+  /// {@template flutter.widgets.restoration.primitivevalue.constructor}
+  /// If no restoration data is available to restore the value in this property
+  /// from, the property will be initialized with the provided `defaultValue`.
+  /// {@endtemplate}
+  RestorableNum(T defaultValue) : super(defaultValue);
+}
+
+/// A [RestorableProperty] that knows how to store and restore a [double].
+///
+/// {@macro flutter.widgets.restoration.primitivevalue}
+class RestorableDouble extends RestorableNum<double> {
+  /// Creates a [RestorableDouble].
+  ///
+  /// {@macro flutter.widgets.restoration.primitivevalue.constructor}
+  RestorableDouble(double defaultValue) : super(defaultValue);
+}
+
+/// A [RestorableProperty] that knows how to store and restore an [int].
+///
+/// {@macro flutter.widgets.restoration.primitivevalue}
+class RestorableInt extends RestorableNum<int> {
+  /// Creates a [RestorableInt].
+  ///
+  /// {@macro flutter.widgets.restoration.primitivevalue.constructor}
+  RestorableInt(int defaultValue) : super(defaultValue);
+}
+
+/// A [RestorableProperty] that knows how to store and restore a [String].
+///
+/// {@macro flutter.widgets.restoration.primitivevalue}
+class RestorableString extends _RestorablePrimitiveValue<String> {
+  /// Creates a [RestorableString].
+  ///
+  /// {@macro flutter.widgets.restoration.primitivevalue.constructor}
+  RestorableString(String defaultValue) : super(defaultValue);
+}
+
+/// A [RestorableProperty] that knows how to store and restore a [bool].
+///
+/// {@macro flutter.widgets.restoration.primitivevalue}
+class RestorableBool extends _RestorablePrimitiveValue<bool> {
+  /// Creates a [RestorableBool].
+  ///
+  /// {@macro flutter.widgets.restoration.primitivevalue.constructor}
+  RestorableBool(bool defaultValue) : super(defaultValue);
+}
+
```

Comment:
do you intend the _RestorablePrimitiveValue subclasses to be for nullable types or non-nullable types?

Reply:
Do document out discussion over chat: I am intending them to be non-nullable since that's gonna be the new default for dart. I added some more asserts to express that intend. If we in the future need a nullable type, we can add one. Those could be named with a trailing N (for Nullable), e.g. RestorableIntN, RestorableStringN, etc. similar to how in lisp functions that return a bool end in p.